File: 0844-backspace-string-compare/0844-backspace-string-compare.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var backspaceCompare = function(s, t) {
    return parseStr(s) === parseStr(t);
};

function parseStr(str) {
    let i = 0, newStr = '';
    while (i < str.length) {
        if (str[i] === '#') {
            newStr = newStr.slice(0,-1);
            i++;
        } else if (str[i+1] === '#') {
            i += 2;
        } else {
            newStr += str[i];
            i++;
        }
    }
    return newStr;
}

File: 0844-backspace-string-compare/0844-backspace-string-compare.py
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
      return self.parse_str(s) == self.parse_str(t)
    
    def parse_str(self, string: str):
      i, stack = 0, []
      while i < len(string):
        if string[i] != '#':
          stack.append(string[i])
        else:
          if len(stack) > 0:
            stack.pop()
        i += 1
      return ''.join(stack)
      